### PR TITLE
Disallow fi_fence reqs to be added to fi_more list

### LINKED ISSUE #1074: FI_MORE doesn't work with FI_FENCE
Right now, an FI_FENCE transaction can be placed on a more list.  FI_FENCE is handled inside _gnix_vc_queue_tx_req().  Putting a FI_FENCE req in a more list will break that implementation.

@a-abraham @sungeunchoi

### PR DESCRIPTION
**Pull request by a-abraham**
fixes ofi-cray/libfabric-cray#1074
Signed-off-by: Amith Abraham <aabraham@cray.com>
@sungeunchoi @ztiffany

### CODE REVIEW

File: prov/gni/src/gnix_rma.c [hunk 1]
```diff
@@ -1441,10 +1446,11 @@ ssize_t _gnix_rma(struct gnix_fid_ep *ep, enum gnix_fab_req_type fr_type,
 	/* Add reads/writes to slist when FI_MORE is present, Or
 	 * if this is the first message in the chain without FI_MORE.
 	 * When FI_MORE is not present, if the slists are not empty
-	 * it is the first message without FI_MORE. */
-	if ((flags & FI_MORE) ||
-	    (!(slist_empty(&ep->more_write)) ||
-	     !(slist_empty(&ep->more_read)))) {
+	 * it is the first message without FI_MORE. Does not add fence
```
Comment (sungeunchoi) [inline prov/gni/src/gnix_rma.c]: Does -> Do

File: prov/gni/src/gnix_rma.c [hunk 2]
```diff
@@ -1346,6 +1346,11 @@ ssize_t _gnix_rma(struct gnix_fid_ep *ep, enum gnix_fab_req_type fr_type,
 		return -FI_EINVAL;
 	}
 
+	/* Disallow fi_more with fi_fence */
```
Comment (sungeunchoi) [inline prov/gni/src/gnix_rma.c]: I think I'd word this differently.  It's really the case that a FI_FENCE breaks the FI_MORE chain.

File: prov/gni/src/gnix_rma.c (hunk 1 above)
Comment (a-abraham) [reply prov/gni/src/gnix_rma.c]: Fixed.

File: prov/gni/src/gnix_rma.c (hunk 2 above)
Comment (a-abraham) [reply prov/gni/src/gnix_rma.c]: Fixed.

**Comment by ztiffany** [comment]
I think we're still missing a piece here.  When using FI_FENCE when there is an existing FI_MORE chain, you will end with initiating the more chain and returning, never calling queue_tx_req() on the new FI_FENCE request.

**Comment by a-abraham** [comment]
@ztiffany  That's a good catch. Though I wonder if I should still remove the FI_MORE flag at the beginning then, might be useful to maintain all that information.

**Comment by a-abraham** [comment]
The following should suffice, I believe.
1) Initiate Chain
- Normally return FI_SUCCESS;  
- If there is an FI_FENCE, return _gnix_vc_queue_tx_req(req);

**Comment by ztiffany** [comment]
👍

File: prov/gni/src/gnix_rma.c [hunk 3]
```diff
@@ -1478,6 +1485,9 @@ ssize_t _gnix_rma(struct gnix_fid_ep *ep, enum gnix_fab_req_type fr_type,
 			_gnix_vc_queue_tx_req(more_req);
 			slist_init(&ep->more_read);
 		}
+		if (flags & FI_FENCE) {
```
Comment (sungeunchoi) [inline prov/gni/src/gnix_rma.c:1490]: Can you add a comment above this explaining why you need it.  FI_FENCE is a bit tricky to follow through here.

Comment (a-abraham) [reply prov/gni/src/gnix_rma.c:1490]: Sounds good - done.

**Comment by sungeunchoi** [comment]
👍

**Pull request merged**